tesseract/consensus/polygon-pos/src/notification.rs

```diff
@@ -1,101 +1,152 @@
+		untrusted_header.header.validators_hash,
+		false,
+	);
+
+	let next_validator_set_hash_match = validate_validator_set_hash(
+		&ValidatorSet::new(trusted_state.next_validators.clone(), None),
+		untrusted_header.header.next_validators_hash,
+		true,
+	);
+
+	let (milestone_number, milestone) = client.prover.get_latest_milestone().await?;
+	let milestone_end_block = milestone.end_block.parse::<u64>().unwrap_or(0);
 
-	if !headers.is_empty() {
-		let chain_head = headers[0].parent_hash;
-		Ok(Some(PolygonClientUpdate {
-			consensus_update: headers.try_into().expect("Infallible"),
-			chain_head,
-		}))
+	let maybe_milestone_update = if milestone_end_block > consensus_state.last_finalized_block {
+		Some(
+			build_milestone_update(
+				client,
+				milestone_number,
+				milestone.clone(),
+				milestone_end_block,
+			)
+			.await?,
+		)
 	} else {
-		Ok(None)
+		None
+	};
+
+	match validator_set_hash_match.is_ok() || next_validator_set_hash_match.is_ok() {
+		true => {
+			let ancestry = client
+				.prover
+				.signed_headers_range(trusted_state.height + 1, latest_height)
+				.await?;
+			let next_validators = client.prover.next_validators(latest_height).await?;
+			return Ok(Some(PolygonConsensusUpdate {
+				tendermint_proof: CodecConsensusProof::from(&ConsensusProof::new(
+					untrusted_header,
+					ancestry,
+					Some(next_validators),
+				)),
+				milestone_update: maybe_milestone_update,
+			}));
+		},
+		false => {
+			// Backward traversal
+			let mut height = latest_height;
+			let mut found = false;
+			let mut matched_header = None;
+			while height > trusted_state.height {
+				let header_res = client.prover.signed_header(height).await;
+				let header = match header_res {
+					Ok(h) => h,
+					Err(_) => {
+						height -= 1;
+						continue;
+					},
+				};
+				let validator_set_hash_match = validate_validator_set_hash(
+					&ValidatorSet::new(trusted_state.validators.clone(), None),
+					header.header.validators_hash,
+					false,
+				);
+				let next_validator_set_hash_match = validate_validator_set_hash(
+					&ValidatorSet::new(trusted_state.next_validators.clone(), None),
+					header.header.next_validators_hash,
+					true,
+				);
+				if validator_set_hash_match.is_ok() || next_validator_set_hash_match.is_ok() {
+					found = true;
+					matched_header = Some(header);
+					break;
+				}
+				height -= 1;
+			}
+			if found {
+				let matched_height = height;
+				let matched_header = matched_header.expect("Header must be present if found");
+				let ancestry = client
+					.prover
+					.signed_headers_range(trusted_state.height + 1, matched_height)
+					.await?;
+				let next_validators = client.prover.next_validators(matched_height).await?;
+				return Ok(Some(PolygonConsensusUpdate {
+					tendermint_proof: CodecConsensusProof::from(&ConsensusProof::new(
+						matched_header,
+						ancestry,
+						Some(next_validators),
+					)),
+					milestone_update: maybe_milestone_update,
```

Comment:
You have to build a new milestone update for `matched_height`

Reply:
Deleted this folder.